File: bin/org/cheminfo/scripting/weka/Classifiers/Weka_SVM.js
/**
 * @object: Weka_SVM
 * Goal: allows creation and usage of a Support Vector Machine classifier.
 *       It works with the Sequential Minimal Optimiation algorithm (SMO)
 * Methods: - createClassifier: creates, parametrizes and trains a classifier
 *          - distributionForInstance: indicates the probability of membership
 *            to each class for an instance
 *          - setOptions: modifies the options of the classifier
 * Usage: First, create, parametrize and train the classifier. then, compute the
 *        probability of membership with distributionForInstance. Option
 *        setting after creation is NOT recommended
 */
 var Weka_SVM = {
    
    
    /**
     * @function: createClassifier
     * Goal: creates, parametrizes and trains a SVM classifier
     * @param options: the options to be set. Available options are:
     *        -D
     *         If set, classifier is run in debug mode and
     *         may output additional info to the console
     *        -no-checks
     *         Turns off all checks - use with caution!
     *         Turning them off assumes that data is purely numeric, doesn't
     *         contain any missing values, and has a nominal class. Turning them
     *         off also means that no header information will be stored if the
     *         machine is linear. Finally, it also assumes that no instance has
     *         a weight equal to 0.
     *         (default: checks on)
     *        -C <double>
     *         The complexity constant C. (default 1)
     *        -N
     *         Whether to 0=normalize/1=standardize/2=neither. (default 0=normalize)
     *        -L <double>
     *         The tolerance parameter. (default 1.0e-3)
     *        -P <double>
     *         The epsilon for round-off error. (default 1.0e-12)
     *        -M
     *         Fit logistic models to SVM outputs. 
     *        -V <double>
     *         The number of folds for the internal
     *         cross-validation. (default -1, use training data)
     *        -W <double>
     *         The random number seed. (default 1)
     *        -K <classname and parameters>
     *         The Kernel to use.
     *         (default: weka.classifiers.functions.supportVector.PolyKernel)
     *
     *        Options specific to kernel weka.classifiers.functions.supportVector.PolyKernel:
     *
     *        -D
     *         Enables debugging output (if available) to be printed.
     *         (default: off)
     *        -no-checks
     *         Turns off all checks - use with caution!
     *         (default: checks on)
     *        -C <num>
     *         The size of the cache (a prime number), 0 for full cache and 
     *         -1 to turn it off.
     *         (default: 250007)
     *        -E <num>
     *         The Exponent to use.
     *         (default: 1.0)
     *        -L
     *         Use lower-order terms.
     *         (default: no)
     * @param train_data: the file containing the training data, its path or the
     *                    Weka data instances
     * @return classifier: returns the created classifier or null if unsuccessful
     */
     createClassifier: function(options, train_data){
         return Weka_Java_SVM.createClassifier(options, train_data);
     },
     

    /**
     * @name: distributionForInstance
     * Goal: computes, for each class, the proportion of membership of given instance
     * @param instance: the instance to be classified
     * @return double[]: the proportion of membership to each class for given instance
     */
     distributionForInstance: function(instance){
         return Weka_Java_SVM.distributionForInstance(instance);
     },


    /**
     * @name: setOptions
     * Goal: modifies the options of the classifier
     * @param options: the options to be set. see createClassifier for available
     *                 options
     * @return boolean: indicates whether the modification was successful or not
     */
     setOptions: function(options){
         return Weka_Java_SVM.setOptions(options);
     }
     
 }

File: src/org/cheminfo/scripting/weka/Classifiers/Weka_SVM.java
/**
 * Name: Weka_SVM
 * Goal: allows creation and usage of a Support Vector Machine classifier.
 *       It works with the Sequential Minimal Optimiation algorithm (SMO)
 * Methods: - createClassifier: creates, parametrizes and trains a classifier
 *          - distributionForInstance: indicates the probability of membership
 *            to each class for an instance
 *          - setOptions: modifies the options of the classifier
 * Usage: First, create, parametrize and train the classifier. then, compute the
 *        probability of membership with distributionForInstance. Option
 *        setting after creation is NOT recommended
 * @author: Numa Trezzini
 */
package org.cheminfo.scripting.weka.Classifiers;

import org.cheminfo.function.Function;
import org.cheminfo.scripting.Utils.DataLoader;
import weka.classifiers.functions.SMO;
import weka.core.Instance;
import weka.core.Instances;

public class Weka_SVM extends Function{

    private SMO classifier;

    /**
     * Name: createClassifier
     * Goal: creates, parametrizes and trains a SVM classifier
     * @param options: the options to be set on the classifier
     * @param train_data: the data used for training. May be the path to the 
     *                    file, the file or data instances
     * @return Classifier: returns the created classifier or null if unsuccessful
     */
    public boolean createClassifier(String options, Object train_data){
        classifier = new SMO();
        try {
            String[] opts = weka.core.Utils.splitOptions(options);
            classifier.setOptions(opts);
        } catch (Exception e) {
            this.appendError(this.getClass().getName(), "Option not recognised: "+e.getMessage());
            System.out.println("option non reconnue:");
            System.out.println(e.getMessage());
            return false;
        }
        
        Instances data = null;
        if(train_data instanceof String)
            data = DataLoader.loadData(null, null, (String)train_data, true);
        else if(train_data instanceof Instances)
            data = (Instances)train_data;
        else {
            System.out.println("train data not supported!");
            this.appendError(this.getClass().getName(), "Training data format not supported");
            return false;
        }
        
        try {
            classifier.buildClassifier(data);
        } catch (Exception e) {
            this.appendError(this.getClass().getName(), "Classifier could not be created: "+e.getMessage());
            System.out.println("le classificateur n'a pas ete cree:");
            System.out.println(e.getMessage());
            return false;
        }
        if(getLogLevel()>2)
            this.appendInfo(this.getClass().getName(), "Classifier was successfully created and trained");
        return true;
    }/*end createClassifier*/
    

    /**
     * Name: distributionForInstance
     * Goal: indicates the proportion of membership to each class for a data 
     *       instance
     * @param c: the classifier to be used
     * @param i: the data instance to be classified
     * @return double[]: indicates, for each class, the porportion of membership
     */
    public double[] distributionForInstance(Instance i){
        try {
            return classifier.distributionForInstance(i);
        } catch (Exception e) {
            this.appendError(this.getClass().getName(), "Instance could not be classified: "+e.getMessage());
            System.out.println("l'instance n'a pas pu etre classee:");
            System.out.println(e.getMessage());
            return null;
        }
    }/*end distributionForInstance*/

    
    /**
     * Name: setOptions
     * Goal: allows modification of the classifiers options
     * @param c: the classifier to be modified
     * @param options: the options to be set
     * @return boolean: indicates whether the modification was successful or not
     */
    public boolean setOptions(String[] options){
        try {
            classifier.setOptions(options);
        } catch (Exception e) {
            this.appendError(this.getClass().getName(), "Option not recognised: "+e.getMessage());
            System.out.println("option non reconnue:");
            System.out.println(e.getMessage());
            return false;
        }
        return true;
    }/*end setOptions*/

}/*end Weka_SVM*/